### PR TITLE
:bug: low-code: Set max_retries value to error handler's

### PR DESCRIPTION
**Pull request by girarda**
## What
The SimpleRetriever inherits the HttpStream's `max_retries` property instead of using the Requester's ErrorHandler's value. This prevents low-code connector developer from configuring how often to retry a failed retryable request

## How
* Override the HttpStream's `max_retries` property
* Delegate its value to the HttpRequester, which delegates it to the ErrorHandler
* Add missing asserts in unit tests

## Testing
I modified the chargebee error handler to test this change:
```
    error_handler:
      type: CompositeErrorHandler
      error_handlers:
        - type: DefaultErrorHandler
          response_filters:
            - http_codes:
                - 404
              action: RETRY
          max_retries: 2
```
and changed the stream's path to an invalid path to force a 404

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py`
2. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py`
3. `airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py`
4. `airbyte-cdk/python/unit_tests/sources/declarative/retrievers/test_simple_retriever.py`
5. `airbyte-cdk/python/unit_tests/sources/declarative/requesters/test_http_requester.py`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Checklist
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Connector version is set to `0.0.1`
    - [ ] `Dockerfile` has version `0.0.1`
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Connector version has been incremented
    - [ ] Version has been bumped according to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines
    - [ ] `Dockerfile` has updated version
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` with an entry for the new version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)

- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def get_request_body_data(
         stream_state: Optional[StreamState] = None,
         stream_slice: Optional[StreamSlice] = None,
         next_page_token: Optional[Mapping[str, Any]] = None,
-    ) -> Optional[Mapping[str, Any]]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py:102]: this method signature is not conform to its parent definition https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_options/request_options_provider.py#L49-L56

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def get_request_body_json(
...
+    ) -> Mapping[str, Any]:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py:120]: this method signature is not conform to its parent definition https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/request_options/request_options_provider.py#L68-L74

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 3]
```diff
@@ -394,6 +394,13 @@ def _parse_records(
     def must_deduplicate_query_params(self) -> bool:
         return True
 
+    @property
+    def max_retries(self) -> Union[int, None]:
+        """
+        Override if needed. Specifies maximum amount of retries for backoff policy. Return None for no limit.
+        """
+        return self.requester.max_retries
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:402]: this is crux of the bug fix

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 4]
```diff
@@ -394,6 +394,13 @@ def _parse_records(
     def must_deduplicate_query_params(self) -> bool:
         return True
 
+    @property
+    def max_retries(self) -> Union[int, None]:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:398]: I'm confused. This only usage I see of this method is in the tests. Can you point out where else it is used?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py [hunk 5]
```diff
@@ -144,3 +144,11 @@ def send_request(
 
         If a log formatter is provided, it's used to log the performed request and response. If it's not provided, no logging is performed.
         """
+
+    @property
+    @abstractmethod
+    def max_retries(self) -> Optional[int]:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/requester.py:150]: How will this behave for other retrievers like the polling one? Will the retries apply on the sets of requests or for each request? 

Having this question makes me feel like the abstraction should be only on the requester level as retries might not be represented in the same way depending on the type of retrieval. For example, the retry for polling might look like:
```
{
  "job_creation": 10,
  "waiting_for_job": 5,
  "fetch_job": 100
}
```

**Review by maxi297 — COMMENTED** [review]
I think I don't get the fix. The PR description says `The SimpleRetriever inherits the HttpStream's max_retries`. However, I don't see the SimpleRetriever inheriting from HttpStream ([SimpleRetriever declaration](https://github.com/airbytehq/airbyte/blob/44f784e200fd66d43e0f948858aa2a2ee184fef7/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L28) and its [parent declaration](https://github.com/airbytehq/airbyte/blob/44f784e200fd66d43e0f948858aa2a2ee184fef7/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/retriever.py#L15)).

I also don't see why `HttpRequester.max_retries` needs to be exposed as it is already used but only internal [here](https://github.com/airbytehq/airbyte/blob/44f784e200fd66d43e0f948858aa2a2ee184fef7/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py#L479). 

So my question would be: what makes you think there is an issue with retries? Using https://github.com/airbytehq/airbyte/compare/maxi297/test_error_handler_max_retries with airbyte_cdk version 0.72.1, I get:

```
{"type": "LOG", "log": {"level": "ERROR", "message": "Giving up _send(...) after 3 tries (airbyte_cdk.sources.streams.http.exceptions.UserDefinedBackoffException: Request URL: https://openexchangerates.org/api/asdfasfasdfasdfasdf/historical/2023-03-13.json?app_id=****&base=USD, Response Code: 405, Response Text: Method Not Allowed)"}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Max retry limit reached. Request: <PreparedRequest [GET]>, Response: <Response [405]>"}}
```

Changing the max_retries to 1, I get a total of two tries:
```
{"type": "LOG", "log": {"level": "ERROR", "message": "Giving up _send(...) after 2 tries (airbyte_cdk.sources.streams.http.exceptions.UserDefinedBackoffException: Request URL: https://openexchangerates.org/api/asdfasfasdfasdfasdf/historical/2023-03-13.json?app_id=****&base=USD, Response Code: 405, Response Text: Method Not Allowed)"}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Max retry limit reached. Request: <PreparedRequest [GET]>, Response: <Response [405]>"}}
```

So what is the issue we are trying to fix? It feels like it's just an old PR that was re-opened

**Comment by natikgadzhi** [comment]
Uh. Is this still a thing?

**Pull request closed without merging**